packages/contracts-bridge/contracts/test/BridgeMessage.t.sol

```diff
@@ -0,0 +1,627 @@
+            .ref(uint40(BridgeMessage.Types.Invalid));
+        assertFalse(BridgeMessage.isValidAction(hookAction));
+    }
+
+    function test_isValidActionRevertWrongActionType() public {
+        bytes29 hookAction = abi.encodePacked(BridgeMessage.Types.Invalid).ref(
+            uint40(BridgeMessage.Types.TransferToHook)
+        );
+        assertFalse(BridgeMessage.isValidAction(hookAction));
+    }
+
+    function test_isValidActionRevertWrongActionTypeViewType() public {
+        bytes29 hookAction = abi.encodePacked(BridgeMessage.Types.Invalid).ref(
+            uint40(BridgeMessage.Types.Invalid)
+        );
+        assertFalse(BridgeMessage.isValidAction(hookAction));
+    }
+
+    /// @notice A BridgeMessage must be at least TOKEN_ID_LEN + MIN_TRANSFER_LEN
+    /// so that it can contain all the required information needed by
+    /// the Bridge. Apart from that, the upper bound is set by the
+    /// Nomad Protocol itself.
+    function test_isValidMessageLength() public {
+        bytes memory longMessage = new bytes(9999999);
+        bytes memory shortMessage = new bytes(10);
+        bytes29 longView = longMessage.ref(0);
+        bytes29 shortView = shortMessage.ref(0);
+        assertFalse(BridgeMessage.isValidMessageLength(shortView));
+        assertTrue(BridgeMessage.isValidMessageLength(longView));
+    }
+
+    function test_formatMessageFailNotAction() public {
+        // I encode the correct type inside the data structure
+        // but set the wrong type in the view
+        // formatMessage() accepts only views of type "Transfer"
+        bytes29 action = abi
+            .encodePacked(
+                BridgeMessage.Types.Transfer,
+                tokenReceiver,
+                tokenAmount,
+                tokenDetailsHash
+            )
+            .ref(uint40(BridgeMessage.Types.Invalid));
+        bytes29 tokenId = BridgeMessage.formatTokenId(
+            localDomain,
+            tokenAddress
+        );
+        vm.expectRevert("!action");
+        BridgeMessage.formatMessage(tokenId, action);
+    }
+
+    function test_formatMessageNotTokenIdType() public {
+        // I encode the correct type inside the data structure
+        // but set the wrong type in the view
+        // formatMessage() accepts only tokenId views of the type "TokenId"
+        bytes29 action = abi
+            .encodePacked(
+                BridgeMessage.Types.Transfer,
+                tokenReceiver,
+                tokenAmount,
+                tokenDetailsHash
+            )
+            .ref(uint40(BridgeMessage.Types.Transfer));
+        bytes29 tokenId = abi.encodePacked(localDomain, tokenAddress).ref(
+            uint40(BridgeMessage.Types.Invalid)
+        );
+        vm.expectRevert(
+            "Type assertion failed. Got 0x0000000000. Expected 0x0000000001"
+        );
+        BridgeMessage.formatMessage(tokenId, action);
+    }
+
+    function test_formatMessageTransfer() public {
```

Comment:
some of this is duplicating the logic. instead we should consider hardcoding a set of inputs, and the correct answer. then checking that the logic produces the correct answer

i.e. to test `add(x, y)`, we can hardcode 1,2 and 3, then invoke `add(1,2)` and check that it returns `3`

this is more of a comment on the purpose of unit tests, for use in designing future tests. no specific change needed right now.

Reply:
So, this is something that occurred to me. How I reasoned about it was that I used irrelevant code paths and followed the simple rule of experimentation. Keep everything stable, change 1 parameter. So how I thought about it was like this: "Assuming all code is correct, except the one I am testing". If some of the logic that I use is incorrect, that should come up in it's unit test?

Not sure if that is the correct way to think about it, just how I approached it.